src/System.Net.Requests/src/System/Net/HttpWebRequest.cs

```diff
@@ -91,6 +95,55 @@ private enum Booleans : uint
             IsWebSocketRequest = 0x00000800,
             Default = AllowAutoRedirect | AllowWriteStreamBuffering | ExpectContinue
         }
+
+        private class HttpClientParameters
+        {
+            public readonly DecompressionMethods AutomaticDecompression;
+
+            public readonly bool AllowAutoRedirect;
+
+            public readonly int MaximumAutomaticRedirections;
+
+            public readonly int MaximumResponseHeadersLength;
+
+            public readonly bool PreAuthenticate;
+
+            public readonly int Timeout;
+
+            public readonly SslProtocols SslProtocols;
+
+            public readonly bool CheckCertificateRevocationList;
+
+            public HttpClientParameters(HttpWebRequest webRequest)
+            {
+                AutomaticDecompression = webRequest.AutomaticDecompression;
+                AllowAutoRedirect = webRequest.AllowAutoRedirect;
+                MaximumAutomaticRedirections = webRequest.MaximumAutomaticRedirections;
+                MaximumResponseHeadersLength = webRequest.MaximumResponseHeadersLength;
+                PreAuthenticate = webRequest.PreAuthenticate;
+                Timeout = webRequest.Timeout;
+                SslProtocols = (SslProtocols)ServicePointManager.SecurityProtocol;
+                CheckCertificateRevocationList = ServicePointManager.CheckCertificateRevocationList;
+            }
+
+            public bool Matches(HttpWebRequest webRequest)
+            {
+                return AutomaticDecompression == webRequest.AutomaticDecompression && AllowAutoRedirect == webRequest.AllowAutoRedirect
+                    && MaximumAutomaticRedirections == webRequest.MaximumAutomaticRedirections
+                    && MaximumResponseHeadersLength == webRequest.MaximumResponseHeadersLength && PreAuthenticate == webRequest.PreAuthenticate
+                    && Timeout == webRequest.Timeout && SslProtocols == (SslProtocols)ServicePointManager.SecurityProtocol
+                    && CheckCertificateRevocationList == ServicePointManager.CheckCertificateRevocationList;
+            }
+
+            public static bool AreParametersAcceptableForCaching(HttpWebRequest webRequest)
+            {
+                return webRequest._credentials == null && ReferenceEquals(webRequest._proxy, DefaultWebProxy)
+                    && webRequest.ServerCertificateValidationCallback == null && ServicePointManager.ServerCertificateValidationCallback == null
+                    && (webRequest._clientCertificates == null || webRequest._clientCertificates.Count == 0)
```

Comment:
We should remove the Count == 0 check.  As the collection is mutable, the collection could have a cert added after this check happens, in which case we'll make the wrong decision, and while it looks like that would require a race condition and misuse to happen, as it's an area of security, we shouldn't take chances.  It also seems like Count == 0 here will be rare: in general I'd expect the collection to either be null or to have > 0 certs in it.  Is that an invalid assumption?

Reply:
> I could add the same concern applies to ServerCertificateCallbackValidation and SslProtocols properties which can be changed after this check is done.

I was about to comment on that as well.  We should be structuring the code such that we read these fields once and then use those values, to compare to the cached instance and to create a new one if necessary.  If the values are concurrently changed on the HttpWebRequest by the developer after we've read them, they'll just be ignored.

Reply:
There seems no problem to re-write the code that way, but it would help only in this specific case. All existing parts the legacy HTTP mechanism will still be concurrently unsafe. Moreover, we cannot freeze WebProxy object here which is not null by default and mutable.
So, this part just follows the legacy approach "set all SPM properties once and don't change them late". Do you think it is still worth solving that concurrency safety problem only in the given isolated case?

Reply:
Regardless, the short answer to your question: yes, I do believe it's worth it here.  We're adding caching, and we're explicitly avoiding caching in cases where security comes into play.  But race conditions here could cause us to incorrectly cache or used cache data with different security policies, and we need to avoid that.

Reply:
I found two places where a field read twice and decision is made based on the first read.

1. [HttpWebRequest - ServerCertificateValidationCallback](https://github.com/dotnet/corefx/blob/dcc9f626eeb0751d4ba4499f623aac42d2e95c10/src/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1161-L1163)
2. [HttpWebRequest - _proxy](https://github.com/dotnet/corefx/blob/dcc9f626eeb0751d4ba4499f623aac42d2e95c10/src/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1144-L1153)

Additionally, [ClientCertificates collection is already passed to AddRange method](https://github.com/dotnet/corefx/blob/dcc9f626eeb0751d4ba4499f623aac42d2e95c10/src/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L1156) which is not thread safe and, in example, can cause IndexOutOfRangeException.

I agree that these operations are not related to caching, however they are performed inside SendRequest method which is marked as thread safe according to our documentation.

Reply:
> however they are performed inside SendRequest method

What is SendRequest?  What docs are you referring to?

> which is marked as thread safe

"Thread safe" doesn't mean there aren't any issues related to threading: it means you can use the method concurrently and things won't get corrupted.

Thanks for the examples. For `HttpWebRequest - ServerCertificateValidationCallback`, this can easily be "fixed" by changing:
```C#
RemoteCertificateValidationCallback rcvc = ServerCertificateValidationCallback != null ?
    ServerCertificateValidationCallback :
    ServicePointManager.ServerCertificateValidationCallback;
```
to:
```C#
RemoteCertificateValidationCallback rcvc = ServerCertificateValidationCallback ??
    ServicePointManager.ServerCertificateValidationCallback;
```
However, even if we didn't change it, this isn't violating thread safety.  ServerCertificateValidationCallback is not documented to be thread-safe, so you shouldn't be changing it concurrently with this operation.  But even if you did, the worst that happens is you change it from non-null to null, in which case due to the race condition we won't use any validation callback.

For `HttpWebRequest - _proxy`, this can also be "fixed" easily by storing the _proxy into a local and then just using that local rather than repeatedly referring to the field, and the worst that happens if you make invalid concurrent modifications while sending the request is you might get an exception like NullReferenceException or a NotSupportedException.

Note that in both of these cases, any such invalid concurrent access only affects that individual request, not other requests.  The issues I highlighted in your PR could end up affecting other completely unrelated requests in the same process.  Whether or not we fix the above cases (which is fine to do), we need to fix the ones in your PR.

Reply:
Sorry, I made a mistake regarding SendRequest. It's not declared as thread safe.
My point is that there are places in the existing code vulnerable to race conditions. However, I agree that caching is more sensitive to this kind of issues and we should avoid double-reads.

Reply:
Fixed.
Parameters are save into a snapshot. Those two double reads are now gone.